lib/babe/epoch.go

```diff
@@ -221,37 +222,69 @@ func (b *Service) incrementEpoch() (uint64, error) {
+) (*types.PreRuntimeDigest, error) {
 	proof, err := claimPrimarySlot(
 		epochData.randomness,
-		slot,
-		epoch,
+		slotNumber,
+		epochNumber,
 		epochData.threshold,
-		b.keypair,
+		keypair,
 	)
-	if err == nil {
-		return types.NewBabePrimaryPreDigest(
-			epochData.authorityIndex, slot, proof.output, proof.proof).ToPreRuntimeDigest()
+
+	switch err {
+	case nil:
+		preRuntimeDigest, err := types.NewBabePrimaryPreDigest(
+			epochData.authorityIndex, slotNumber, proof.output, proof.proof).ToPreRuntimeDigest()
+		if err != nil {
+			return nil, fmt.Errorf("error converting babe primary pre-digest to pre-runtime digest: %w", err)
+		}
+		logger.Debugf("epoch %d: claimed primary slot %d", epochNumber, slotNumber)
+		return preRuntimeDigest, nil
+	default:
+		if !errors.Is(err, errOverPrimarySlotThreshold) {
+			return nil, fmt.Errorf("error running slot lottery at slot %d: %w", slotNumber, err)
+		}
 	}
 
-	if errors.Is(err, errOverPrimarySlotThreshold) {
-		if epochData.secondary == 0 {
-			return nil, errSecondarySlotProductionDisabled
+	switch epochData.secondary {
+	case types.PrimarySlots:
+		return nil, errSecondarySlotProductionDisabled
+	case types.PrimaryAndSecondaryVRFSlots:
+		proof, err := claimSecondarySlotVRF(
+			epochData.randomness, slotNumber, epochNumber, epochData.authorities, keypair, epochData.authorityIndex)
+		if err != nil {
+			return nil, fmt.Errorf("error claim secondary vrf slot at %d: %w", slotNumber, err)
+		}
+		preRuntimeDigest, err := types.NewBabeSecondaryVRFPreDigest(
+			epochData.authorityIndex, slotNumber, proof.output, proof.proof).ToPreRuntimeDigest()
+		if err != nil {
+			return nil, fmt.Errorf("error converting babe secondary vrf pre-digest to pre-runtime digest: %w", err)
+		}
+		logger.Debugf("epoch %d: claimed secondary vrf slot %d", epochNumber, slotNumber)
+		return preRuntimeDigest, nil
+	case types.PrimaryAndSecondaryPlainSlots:
+		err = claimSecondarySlotPlain(
+			epochData.randomness, slotNumber, epochData.authorities, epochData.authorityIndex)
+		if err != nil {
+			return nil, fmt.Errorf("error claiming secondary plain slot at %d: %w", slotNumber, err)
 		}
 
-		err = claimSecondarySlotPlain(epochData.randomness, slot,
-			epochData.authorities, epochData.authorityIndex)
+		preRuntimeDigest, err := types.NewBabeSecondaryPlainPreDigest(
+			epochData.authorityIndex, slotNumber).ToPreRuntimeDigest()
 		if err != nil {
-			return nil, fmt.Errorf("cannot claim secondary slot: %w", err)
+			return nil, fmt.Errorf(
+				"failed to get preruntime digest from babe secondary plain predigest for slot %d: %w", slotNumber, err)
 		}
-		return types.NewBabeSecondaryPlainPreDigest(epochData.authorityIndex, slot).ToPreRuntimeDigest()
+		logger.Debugf("epoch %d: claimed secondary plain slot %d", epochNumber, slotNumber)
+		return preRuntimeDigest, nil
 	}
 
-	return nil, err
+	return nil, errors.New("invalid slot claiming technique")
```

Comment:
define a sentinel error

Reply:
do you mean defining a variable for this? This error should never occur, I think we don't have to